src/lightgbm/src/main/scala/TrainUtils.scala

```diff
@@ -0,0 +1,117 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in project root for information.
+
+package com.microsoft.ml.spark
+
+import com.microsoft.ml.lightgbm.{SWIGTYPE_p_void, lightgbmlib, lightgbmlibConstants}
+import org.apache.spark.TaskContext
+import org.apache.spark.ml.linalg.{DenseVector, SparseVector}
+import org.apache.spark.sql.Row
+import org.slf4j.Logger
+
+case class TrainParams(numIterations: Int, learningRate: Double, numLeaves: Int)
+
+private object TrainUtils extends java.io.Serializable {
+  private val DefaultBufferLength: Int = 10000
+
+  def translate(labelColumn: String, featuresColumn: String, parallelism: String, log: Logger,
+                trainParams: TrainParams, inputRows: Iterator[Row]): Iterator[LightGBMBooster] = {
+    if (!inputRows.hasNext)
+      List[LightGBMBooster]().toIterator
+
+    val rows = inputRows.toArray
+    val numRows = rows.length
+    val rowsAsDoubleArray = rows.map(row => (row.get(row.fieldIndex(featuresColumn)) match {
+      case dense: DenseVector => dense.toArray
+      case sparse: SparseVector => sparse.toDense.toArray
```

Comment:
It seems strange that you are converting sparse features to dense. LightGBM supports sparse features natively, see:  LGBM_DatasetCreateFromCSR and  LGBM_DatasetCreateFromCSC. Making a dense matrix from sparse rows will be very memory inefficient.

Reply:
yes, I was planning to add sparse support later, but I can add it now.
In the case where there are mixed sparse and dense vectors I will just convert the rest to the format of the first row, as an intelligent computation may be expensive.